include/rsp.h

```diff
@@ -10,12 +10,206 @@
+#define SP_WSTATUS_CLEAR_SIG3        0x08000
+#define SP_WSTATUS_SET_SIG3          0x10000
+#define SP_WSTATUS_CLEAR_SIG4        0x20000
+#define SP_WSTATUS_SET_SIG4          0x40000
+#define SP_WSTATUS_CLEAR_SIG5        0x80000
+#define SP_WSTATUS_SET_SIG5          0x100000
+#define SP_WSTATUS_CLEAR_SIG6        0x200000
+#define SP_WSTATUS_SET_SIG6          0x400000
+#define SP_WSTATUS_CLEAR_SIG7        0x800000
+#define SP_WSTATUS_SET_SIG7          0x1000000
+
+/**
+ * @brief RSP ucode definition.
+ * 
+ * This small structure holds the text/data pointers to a RSP ucode program
+ * in RDRAM. It also contains the name (for the debugging purposes) and
+ * the initial PC (usually 0).
+ * 
+ * If you're using libdragon's build system (n64.mk), use DEFINE_RSP_UCODE()
+ * to initialize one of these.
+ */
+typedef struct {
+    uint8_t *code; void *code_end;
+    uint8_t *data; void *data_end;
+    const char *name; uint32_t start_pc;
+} rsp_ucode_t;
+
+/**
+ * @brief Define one RSP ucode compiled via libdragon's build system (n64.mk).
+ * 
+ * If you're using libdragon's build system (n64.mk), use DEFINE_RSP_UCODE() to
+ * define one ucode coming from a rsp_*.S file. For instance, if you wrote
+ * and compiled a ucode called rsp_math.S, you can use DEFINE_RSP_UCODE(rsp_math)
+ * to define it at the global level. You can then use rsp_load(&rsp_math) 
+ * to load it.
+ */
+#define DEFINE_RSP_UCODE(ucode_name) \
+    extern uint8_t ucode_name ## _text_start[]; \
+    extern uint8_t ucode_name ## _data_start[]; \
+    extern uint8_t ucode_name ## _text_end[0]; \
+    extern uint8_t ucode_name ## _data_end[0]; \
+    rsp_ucode_t ucode_name = (rsp_ucode_t){ \
+        .code = ucode_name ## _text_start, \
+        .data = ucode_name ## _data_start, \
+        .code_end = ucode_name ## _text_end, \
+        .data_end = ucode_name ## _data_end, \
+        .name = #ucode_name, .start_pc = 0, \
+    }
+
+/** @brief Initialize the RSP subsytem. */
+void rsp_init(void);
+
+/** @brief Load a RSP ucode.
+ * 
+ * This function allows to load a RSP ucode into the RSP internal memory.
+ * The function executes the transfer right away, so it is responsibility
+ * of the caller making sure that it's a good time to do it.
+ * 
+ * The function internally keeps a pointer to the last loaded ucode. If the
+ * ucode passed is the same, it does nothing. This makes it easier to write
+ * code that optimistically switches between different ucodes, but without
+ * forcing transfers every time.
+ * 
+ * @param[in]     ucode       Ucode to load into RSP
+ **/
+void rsp_load(rsp_ucode_t *ucode);
+
+/** @brief Run RSP ucode.
+ * 
+ * This function starts running the RSP, and wait until the ucode is finished.
+ */
+void rsp_run(void);
+
+/** @brief Run RSP async.
+ * 
+ * This function starts running the RSP in background. Use rsp_wait() to
+ * synchronize later.
+ */
+void rsp_run_async(void);
+
+/** @brief Wait until RSP is idle. */
+void rsp_wait(void);
```

Comment:
what's the difference? `break` is the way you halt the processor from within the RSP, and the only thing it makes sense to wait is for the processor to be halted. Maybe the comment is wrong?

I think what we're going to have, at some point, is more granularity like "enqueue one task and wait for that task to be done", but since there's no API surface about this now, I think it's premature.

Reply:
Breaking the rsp triggers the SP interrupt, so we can also wait for that when we have the kernel for example, or with a callback etc. In that case I think we will just change the implementation then? I also think its early to decide and that was the reason I wanted to reserve the name by making it more descriptive. Nothing I will insist at all though, if you think we can change the implementation without breaking assumptions of usages won't be a problem.

Reply:
Yes if we implement a interrupt-driven queue, `rsp_wait()` could be changed to follow that. I think the semantic is clear. Maybe the comment should say "until the RSP has finished processing" which is more "general".

For kernelizing this function (and other similar), my plan is as follows. Current implementation:

```C
while (!(*SP_STATUS & SP_STATUS_HALTED)) {} 
```

Future implementation:

```C
while (!(*SP_STATUS & SP_STATUS_HALTED)) {   
    kthread_wait_event(&KEVENT_IRQ_SP);
} 
```

with `kthread_wait_event()` being a no-op when the kernel is not initialized.